Embed Podman and drop Docker support

apyrgio opened the pull request:
This PR introduces a very core change to Dangerzone, that is required for the Independent Container Updates feature (https://github.com/freedomofpress/dangerzone/issues/1006). It adds first class support for Podman, by embedding it into our Windows / macOS packages, and using it internally to create VMs in these platforms where we can run containers. As a result, we drop Docker support, since it's no longer needed, and because it's not compatible with the format of the image we use in the Independent Containers Update feature.

This PR also introduces some visual changes. First, it allows users to select documents without waiting for the container installation to finish, or the Podman machine to start. We have measured that after the first time that Dangerzone runs, where initialization takes a while, the rest of the times require a few seconds to get Dangerzone starting, so we figued it's fine to let the user select documents, and do some work in the background. Second, we now have a widget that shows all the logs of our background tasks, which will help when debugging startup issues.

Finally, the integration of Podman allows us to do end-to-end CI tests on Windows and macOS platforms, which was previously highly difficult. This means that each PR, and more critically all of our Podman machine logic, will be tested on the two main platforms we are supporting, and we will catch errors sooner.

Here is a list of open issues that this PR affects:
* Closes #1170
* Closes #1172
* Refs containers/podman-py#558
* Closes #1168
* Fixes #118
* Fixes #1045
* Fixes #1145
* Fixes #1171
* Fixes #1184
* Closes #253
* Closes #470
* Closes #1132
* Fixes #1009
* Closes #668

apyrgio commented on dangerzone/podman/command/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from .command import PodmanCommand
```
I can create an issue for more context, and add an anchor to that issue here.

apyrgio commented on dangerzone/podman/command/cli_runner.py:
```diff
@@ -0,0 +1,268 @@
+    """Global options for Podman commands.
+
+    Attributes:
+        cdi_spec_dir (Union[str, Path, list[str], list[Path], None]): The CDI spec directory path (can be a list of paths).
```
Yeah, I was trying to copy the way they were doing docstrings, and I recall they include the types of the arguments as well. I haven't done it for all arguments though, because I wanted some feedback from them first. Hence this half-baked docstring :-/

apyrgio commented on dangerzone/podman/command/cli_runner.py:
```diff
@@ -0,0 +1,268 @@
+
+    def display(self, cmd):
+        """Display a list of command-line options as a single command invocation."""
+        parts = [str(part) for part in cmd]
```
Sure, looks tight and I'll change it. My go-to tool is list comprehensions, but I'm not in the camp of banning `map()` as some other folks.

My only concern is if it's worth making trivial changes like this for style reasons, or if there's a stronger reason to do so. We may find various similar occurrences in this PR, and I'm sure it's worth replacing all of them :-/

apyrgio commented on dangerzone/container_utils.py:
```diff
@@ -10,10 +10,18 @@
 
 # Keep the name of the old container here to be able to get rid of it later
 OLD_CONTAINER_NAME = "dangerzone.rocks/dangerzone"
+CONTAINERS_CONF_PATH = get_cache_dir() / "containers.conf"
```
Well, the problem was that I wanted `init_podman_command()` to be in the `dangerzone.container_utils` module, since it it seemed to me as something Podman-related, but not necessarily Podman **machine** -related. So, if I moved just the Podman machine literals there and keep the rest as is, this would create a cyclic dependency.

I could probably rename the `podman.machine` module to something more broad, and then move these functions there. How do you feel about this?

apyrgio replied:
Yeah, that what I mean with renaming `podman.machine` to something more broad. Ok, I'll think of something.

apyrgio commented on dangerzone/podman/cli.py:
```diff
@@ -0,0 +1,141 @@
+        else:
+            click.echo("No Dangerzone Podman machines found.")
+    except PodmanError as e:
+        click.echo(click.style(f"Error: {e}", fg="red"), err=True)
```
Hm, really? I'm not seeing it in the code actually 🤔

almet replied:
Here is one example in the `dangerzone.updater.cli` module

https://github.com/freedomofpress/dangerzone/blob/5bc25e8d2a5a9f9fbd35583b630109b97a9e513d/dangerzone/updater/cli.py#L53-L55

apyrgio commented on dangerzone/podman/cli.py:
```diff
@@ -0,0 +1,141 @@
+@main.command()
+@click.option("--cpus", type=int, help="Number of CPUs to allocate.")
+@click.option("--memory", type=int, help="Amount of memory in bytes.")
+@click.option("--timezone", type=str, help="Timezone for the machine.")
```
The default is `None` here, which means that Podman's defaults will be used. These may defer per OS or change in next versions, so I didn't define them here.

almet replied:
Sorry, that's on me I thought these options where arguments.

apyrgio commented on dev_scripts/dangerzone-machine:
```diff
@@ -0,0 +1,13 @@
+import sys
+
+# Load dangerzone module and resources from the source code tree
+sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
Sure. This is the snippet we copy-paste across the rest of the scripts, so I guess I'll change it everywhere.

apyrgio replied:
Fixed in 06d84f21

apyrgio commented on dev_scripts/dangerzone-machine:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env python3
```
Sigh, probably we could do something better here, but it was only in the backburner :-/

apyrgio commented on dangerzone/gui/log_window.py:
```diff
@@ -0,0 +1,80 @@
+
+    def emit(self, record):
+        msg = self.format(record)
+        self.new_record.emit(msg + "\n")
```
The receiver is `TracebackWidget.process_output()` which is used by other places as well and doesn't add a newline. Given that we handle a log message here, I think it's the responsibility of the log handler to translate it as "this message should be followed by a new line".

apyrgio commented on dangerzone/startup.py:
```diff
@@ -0,0 +1,166 @@
+        logger.info(f"There is an update for the Dangerzone sandbox")
+
+
+class StartupLogic:
```
You could take a look at `dangerzone.gui.startup`, since we subclass it there.

almet commented on dangerzone/container_utils.py:
```diff
@@ -183,6 +193,43 @@ def make_seccomp_json_accessible(runtime: Runtime) -> Union[Path, PurePosixPath]
+helper_binaries_dir=["{helper_binaries_dir}"]
+"""
+    # FIXME: Do not unconditionally write to this file.
+    dst = CONTAINERS_CONF_PATH
```
heh, 3 times ;-) Don't block on it of course :-)

apyrgio commented on dangerzone/podman/command/cli_runner.py:
```diff
@@ -0,0 +1,268 @@
+            path = shutil.which("podman")
+            if path is None:
+                raise errors.PodmanNotInstalled()
+            Path(path)
```
Fixed in 76a6e1bc

apyrgio commented on dangerzone/gui/main_window.py:
```diff
@@ -120,6 +123,73 @@ def get_supported_extensions() -> List[str]:
+        # XXX: This trick works only for SVGs that use `animateTransform`. The rest of
+        # the SVGs that use a different `animate*` property will NOT be animated.
+        self.spinner = QSvgWidget()
+        self.spinner.renderer().setFramesPerSecond(60)
```
Ah, ok. What's the downside of 60FPS though, that it will use more CPU?

apyrgio replied:
What. The.

I didn't think the performance penalty would be too much, but I trust your instinct and I figured, what the heck, let's check it out. I started the Dangerzone GUI, and with `pdb` I set the FPS to 0, 1, 10, 20, 60, and 600. Turns out that from 60FPS onward, there's a significant performance penalty (~2% CPU overhead with 60FPS and ~12% CPU overhead with 600FPS). With 20FPS, there's a penalty still, but it's minimum (less than 0.5%).

So, thanks for the suggestion, I'll set it to 20FPS then. I've also tested that once the spinner is no longer shown, there's no rendering involved and we don't have any penalty.

apyrgio replied:
Fixed in aca4de79e3170de4e7c80f760d8b4310ba84996a.

apyrgio commented on dangerzone/gui/main_window.py:
```diff
@@ -265,25 +269,11 @@ def __init__(self, dangerzone: DangerzoneGui) -> None:
-            self.waiting_widget.show()
-            self.content_widget.hide()
+        self.waiting_widget = WaitingWidget()
+        self.content_widget.hide()
```
Fixed in ed5433b4.

almet commented on dangerzone/podman/command/command.py:
```diff
@@ -0,0 +1,208 @@
+        cmd = self.runner.construct("system", "service", uri, time=time, cors=cors)
+        return self.runner.run_raw(cmd, wait=False, **skwargs)
+
+    def stop_service(
```
If we do not need to use the returned `proc`, then yes, I believe that would be easier to handle it ourselves rather than pass it along.

apyrgio replied:
Cool, cool, I have done a rewrite in 001b6667bd2c15a0bde633c0d77ba03624d867f1 (not tested yet, because this module is broken after all)

almet commented:
11. `dangerzone-machine remove` command is not working

When trying to remove the machine with `poetry run ./dev_scripts/dangerzone-machine remove`, I see the following error message:

```
Are you sure you want to remove machine 'dz-internal-0.9.1'? [y/N]: y
INFO:dangerzone.podman.machine:Removing Podman machine: dz-internal-0.9.1
❌ The Podman process failed with the following error: Command '['/opt/podman/bin/podman', 'machine', 'rm', 'dz-internal-0.9.1']' returned non-zero exit status 125.
Stdout: b'The following files will be deleted:\n\n\n/Users/alexis/.config/containers/podman/machine/applehv/dz-internal-0.9.1.json\n/var/folders/88/3q7jndss7yl_jsvlfmjh12z00000gn/T/podman/dz-internal-0.9.1.sock\n/var/folders/88/3q7jndss7yl_jsvlfmjh12z00000gn/T/podman/dz-internal-0.9.1-gvproxy.sock\n/var/folders/88/3q7jndss7yl_jsvlfmjh12z00000gn/T/podman/dz-internal-0.9.1-api.sock\n/var/folders/88/3q7jndss7yl_jsvlfmjh12z00000gn/T/podman/dz-internal-0.9.1.log\nAre you sure you want to continue? [y/N] '
Stderr: b'Error: EOF\n'
Aborted!
```

almet commented on dangerzone/podman/cli.py:
```diff
@@ -0,0 +1,137 @@
+
+@main.command()
+@click.option("-f", "--force", is_flag=True, help="Force removal without prompt.")
+def remove(force: bool) -> None:
```
`dangerzone-machine remove` command is not working for me:

When trying to remove the machine with `poetry run ./dev_scripts/dangerzone-machine remove`, I see the following error message:

```
Are you sure you want to remove machine 'dz-internal-0.9.1'? [y/N]: y
INFO:dangerzone.podman.machine:Removing Podman machine: dz-internal-0.9.1
❌ The Podman process failed with the following error: Command '['/opt/podman/bin/podman', 'machine', 'rm', 'dz-internal-0.9.1']' returned non-zero exit status 125.
Stdout: b'The following files will be deleted:\n\n\n/Users/alexis/.config/containers/podman/machine/applehv/dz-internal-0.9.1.json\n/var/folders/88/3q7jndss7yl_jsvlfmjh12z00000gn/T/podman/dz-internal-0.9.1.sock\n/var/folders/88/3q7jndss7yl_jsvlfmjh12z00000gn/T/podman/dz-internal-0.9.1-gvproxy.sock\n/var/folders/88/3q7jndss7yl_jsvlfmjh12z00000gn/T/podman/dz-internal-0.9.1-api.sock\n/var/folders/88/3q7jndss7yl_jsvlfmjh12z00000gn/T/podman/dz-internal-0.9.1.log\nAre you sure you want to continue? [y/N] '
Stderr: b'Error: EOF\n'
Aborted!
```

almet commented on dangerzone/podman/machine.py:
```diff
@@ -0,0 +1,124 @@
+logger = logging.getLogger(__name__)
+
+
+class PodmanMachineManager:
```
Interestingly, this might be related to the conflict between the two ways to interact with podman machines.

After resetting the machines with `poetry run ./dev_scripts/dangerzone-machine reset && poetry run ./dev_scripts/dangerzone-machine init`, I got it to work, and was able to do a conversion.

The thing is... when I started back Podman Dekstop, the previous default machine: `podman-machine-default` is now gone.

If we can find a way to isolate further the dangerzone installation from the default Podman Desktop experience, I think that would be valuable.

apyrgio commented:
> 1. Detection that podman is not available locally and needs to be installed.

Fair point, we need to craft the message in a way that we mention `mazette`, but also be useful to end-users, in case their installation is borked somehow.

> 2. Update the colors used in the status bar text
>
> At least when on dark mode, they aren't great, see the screenshot here. The red and green should be updated from my point of view. Using https://webaim.org/resources/contrastchecker/ can help ensure the contrast is good enough.
>
> ![](https://private-user-images.githubusercontent.com/67073/484564630-363ed1e0-3b0c-43ba-829f-857f23b973fc.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4ODc4NDcsIm5iZiI6MTc1Njg4NzU0NywicGF0aCI6Ii82NzA3My80ODQ1NjQ2MzAtMzYzZWQxZTAtM2IwYy00M2JhLTgyOWYtODU3ZjIzYjk3M2ZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTAzVDA4MTkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5NDE4MzFiOTA0ZTA3ZDI4OGM5NDQ3YzZiZmExNzY5MDM0NjNjY2RkY2ZlYjUwYTY2MjY3NGIwNzA3MTExNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jAjgq6cLZzwChachGquBnzQ1lDbBm26fkptJBYM1fls)

Damn, I didn't know this. Does this mean we need different GIF as well? Also, do we need some soft of palette project-wide for consistency? Pinging @harrislapiroff  for this as well.

> 3. Error messages are duplicated, maybe due to how exceptions are handled
>
> Maybe not directly related to your work, but I believe it's worth fixing here, see the screenshot:
>
> ![](https://private-user-images.githubusercontent.com/67073/484565411-34e2a263-1cca-4de2-9d5f-e3ccf54107da.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4ODc4NDcsIm5iZiI6MTc1Njg4NzU0NywicGF0aCI6Ii82NzA3My80ODQ1NjU0MTEtMzRlMmEyNjMtMWNjYS00ZGUyLTlkNWYtZTNjY2Y1NDEwN2RhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTAzVDA4MTkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkMGU0ZDJmNjY1NTIxYzA3YmFlYmUxYmNlMDg1NDY1ZWVmNmFhZDdhYmJlYjExNWI3Yjk3ZDUzZGQ0Y2I2MzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WX541Sa2izVA-AdsTdl9eCXpVDLqiC0rAdd3fEXObjA)

It's directly related to this PR. I'm currently logging errors in the task level, and the startup manager level. This is just in case something we lose an exception in any of those two levels, but in the common path, we end up duplicating them. I think it's something we can fix later on, unless you have a smooth way to log once.

> 4. When installing, I didn't see any wheel.
>
> I'm not sure if it's because of the night mode or because of something else, but that was missing to me :-)

Hm, the spinner looks like this:

![spinner](https://raw.githubusercontent.com/freedomofpress/dangerzone/abad53abb69627a9bc7d7e50c00a8070fb1e1c96/share/spinner.svg). It's kinda grey, so I think it classes with dark mode. I think we need a white variant as well.

> 5. Update the dangerzone-image script to use the embedded podman

This is what should happen behind the scenes, so let's figure out what's going on here.

> 6. Make it easier to download the signed container.tar archive

We really need to publish our first container officially. Once we have this, it will work as a bootstrap.

> 7. When debugging, knowning what the commands are can be useful

Yup, we have discussed this need before. I'm capturing it here (https://github.com/freedomofpress/dangerzone/issues/1236) and we can expand on it.

apyrgio commented on dangerzone/podman/cli.py:
```diff
@@ -0,0 +1,137 @@
+
+
+@main.command()
+def start() -> None:
```
Hm, given that `dangerzone-machine` is a troubleshooting command, I think it's important to get that output. For instance, I have managed to create a Podman machine in rootful mode in the past, so such info would be helpful.

almet replied:
From my perspective, I found this potentially misleading for the end user, as they might not know what a rootless container is, and the message from podman is clearly indicating that this could lead to issues.

If that's a troubleshooting command intended to us help debug, then it makes sense to get the unfiltered output.

I get what you're saying, and because I don't have an easy-to-implement solution for this, I'll not die on this hill :-)

apyrgio commented on dangerzone/gui/main_window.py:
```diff
@@ -121,6 +152,67 @@ def get_supported_extensions() -> List[str]:
+        layout.addWidget(self.spinner)
+        layout.addWidget(self.message)
+        layout.addWidget(self.info_icon)
+        self.setLayout(layout)
```
Related to https://github.com/freedomofpress/dangerzone/issues/799. I can't reproduce this locally, but if you can, we can look at our CSS and figure something out.

apyrgio commented:
Hey @almet. I believe I've addressed all your comments, both the ones you mentioned here, and some we had discussed live. I'm afraid I don't have the time right now to pinpoint the discussions above to specific commits, but if you take a look at the commits from d3c4dcb7 and onward, you'll see the changes there. Thanks again for your input :slightly_smiling_face:

almet commented on dangerzone/gui/log_window.py:
```diff
@@ -0,0 +1,109 @@
+
+    def handle_task_container_install(self) -> None:
+        self.label.setText(
+            "Installing the Dangerzone container sandbox<br>"
```
```suggestion
            "Installing the Dangerzone sandbox<br>"
```

apyrgio commented on dangerzone/podman/command/command.py:
```diff
@@ -0,0 +1,225 @@
+        self,
+        timeout: Optional[int] = None,
+    ) -> int:
+        """Stop the Podman system service.
```
You're right, I've been thinking about this as well. I was actually inclined to issue the `podman machine stop` command, and then exit, without waiting in the background. After all, the app can work with a Podman machine that's already running, and we can't do much if the Podman machine cannot stop. What do you think?

apyrgio replied:
Hm, while it does work, I'm afraid it's a bit racy. If you start Dangerzone, close it immediately, and then start it again, I've seen the stop action taking place after the start action, which causes an error. I'll check if a message in the status bar seems better.

almet commented on dangerzone/gui/main_window.py:
```diff
@@ -121,6 +160,87 @@ def get_supported_extensions() -> List[str]:
+        self.set_status_working("Initializing Dangerzone VM")
+
+    def handle_task_machine_start(self) -> None:
+        self.set_status_working("Starting Dangerzone VM")
```
Yep, that sounds 'bout right!

apyrgio commented on dangerzone/gui/log_window.py:
```diff
@@ -0,0 +1,109 @@
+        self.label.setText("Checking for updates… failed")
+
+    def handle_task_container_install_failed(self, error: str) -> None:
+        self.label.setText("Installing container sandbox… failed")
```
Fixed in c9f0fa2294b0473f8ddf331f95536b74545653e9

almet commented:
Just got this exception when trying this branch on my mac m1 machine, after starting dangerzone with Podman Desktop and the default podman machine already started.

I'm not sure why but the connection is refused:

```
[DEBUG] Inferred system color scheme as OSColorMode.DARK
qt.qpa.drawing: Layer-backing is always enabled.  QT_MAC_WANTS_LAYER/_q_mac_wantsLayer has no effect.
qt.qpa.drawing: Layer-backing is always enabled.  QT_MAC_WANTS_LAYER/_q_mac_wantsLayer has no effect.
qt.qpa.fonts: Populating font family aliases took 75 ms. Replace uses of missing font family "Consolas" with one that exists to avoid this cost.
[DEBUG] Starting Dangerzone background tasks
qt.qpa.drawing: Layer-backing is always enabled.  QT_MAC_WANTS_LAYER/_q_mac_wantsLayer has no effect.
[INFO] Performing some Dangerzone startup tasks
[DEBUG] /opt/podman/bin/podman machine list --format json
[DEBUG] [
    {
        "Name": "dz-internal-0.9.1",
        "Default": false,
        "Created": "2025-09-03T16:03:59.668536+02:00",
        "Running": false,
        "Starting": false,
        "LastUp": "2025-09-24T20:24:01.526939+02:00",
        "Stream": "default",
        "VMType": "applehv",
        "CPUs": 4,
        "Memory": "2147483648",
        "DiskSize": "107374182400",
        "Port": 62945,
        "RemoteUsername": "core",
        "IdentityPath": "/Users/alexis/.local/share/containers/podman/machine/machine",
        "UserModeNetworking": true
    },
    {
        "Name": "podman-machine-default",
        "Default": false,
        "Created": "2025-09-24T15:37:22.829681+02:00",
        "Running": false,
        "Starting": false,
        "LastUp": "2025-09-24T20:15:19.755979+02:00",
        "Stream": "default",
        "VMType": "applehv",
        "CPUs": 4,
        "Memory": "3999268864",
        "DiskSize": "22548578304",
        "Port": 64148,
        "RemoteUsername": "core",
        "IdentityPath": "/Users/alexis/.local/share/containers/podman/machine/machine",
        "UserModeNetworking": true
    }
]
[DEBUG]
[INFO] Task 'Stopping other Podman VMs' will be skipped
[INFO] Task 'Initializing Dangerzone VM' is starting...
[DEBUG] /opt/podman/bin/podman machine list --format json
[DEBUG] [
    {
        "Name": "dz-internal-0.9.1",
        "Default": false,
        "Created": "2025-09-03T16:03:59.668536+02:00",
        "Running": false,
        "Starting": false,
        "LastUp": "2025-09-24T20:24:01.526939+02:00",
        "Stream": "default",
        "VMType": "applehv",
        "CPUs": 4,
        "Memory": "2147483648",
        "DiskSize": "107374182400",
        "Port": 62945,
        "RemoteUsername": "core",
        "IdentityPath": "/Users/alexis/.local/share/containers/podman/machine/machine",
        "UserModeNetworking": true
    },
    {
        "Name": "podman-machine-default",
        "Default": false,
        "Created": "2025-09-24T15:37:22.829681+02:00",
        "Running": false,
        "Starting": false,
        "LastUp": "2025-09-24T20:15:19.755979+02:00",
        "Stream": "default",
        "VMType": "applehv",
        "CPUs": 4,
        "Memory": "3999268864",
        "DiskSize": "22548578304",
        "Port": 64148,
        "RemoteUsername": "core",
        "IdentityPath": "/Users/alexis/.local/share/containers/podman/machine/machine",
        "UserModeNetworking": true
    }
]
[DEBUG]
[INFO] Podman machine 'dz-internal-0.9.1' already exists.
[INFO] Task 'Initializing Dangerzone VM' completed successfully!
[INFO] Task 'Starting Dangerzone VM' is starting...
[DEBUG] /opt/podman/bin/podman machine list --format json
[DEBUG] [
    {
        "Name": "dz-internal-0.9.1",
        "Default": false,
        "Created": "2025-09-03T16:03:59.668536+02:00",
        "Running": false,
        "Starting": false,
        "LastUp": "2025-09-24T20:24:01.526939+02:00",
        "Stream": "default",
        "VMType": "applehv",
        "CPUs": 4,
        "Memory": "2147483648",
        "DiskSize": "107374182400",
        "Port": 62945,
        "RemoteUsername": "core",
        "IdentityPath": "/Users/alexis/.local/share/containers/podman/machine/machine",
        "UserModeNetworking": true
    },
    {
        "Name": "podman-machine-default",
        "Default": false,
        "Created": "2025-09-24T15:37:22.829681+02:00",
        "Running": false,
        "Starting": false,
        "LastUp": "2025-09-24T20:15:19.755979+02:00",
        "Stream": "default",
        "VMType": "applehv",
        "CPUs": 4,
        "Memory": "3999268864",
        "DiskSize": "22548578304",
        "Port": 64148,
        "RemoteUsername": "core",
        "IdentityPath": "/Users/alexis/.local/share/containers/podman/machine/machine",
        "UserModeNetworking": true
    }
]
[DEBUG]
[INFO] Starting Podman machine: dz-internal-0.9.1
[DEBUG] /opt/podman/bin/podman machine start dz-internal-0.9.1
[DEBUG] Starting machine "dz-internal-0.9.1"

2025-09-24 20:24:08.872 Python[5876:24538862] +[IMKClient subclass]: chose IMKClient_Modern
2025-09-24 20:24:08.872 Python[5876:24538862] +[IMKInputSession subclass]: chose IMKInputSession_Modern
[DEBUG]

[DEBUG] This machine is currently configured in rootless mode. If your containers

[DEBUG] require root permissions (e.g. ports < 1024), or if you run into compatibility

[DEBUG] issues with non-podman clients, you can switch using the following command:

[DEBUG]

[DEBUG] 	podman machine set --rootful dz-internal-0.9.1

[DEBUG]

[DEBUG] API forwarding listening on: /var/run/docker.sock

[DEBUG] Docker API clients default to this address. You do not need to set DOCKER_HOST.

[DEBUG]

[DEBUG] Machine "dz-internal-0.9.1" started successfully

[INFO] Podman machine 'dz-internal-0.9.1' started successfully.
[INFO] Task 'Starting Dangerzone VM' completed successfully!
[INFO] Task 'Check for updates' is starting...
[DEBUG] Cooling down update checks
[INFO] Task 'Check for updates' completed successfully!
[DEBUG] /opt/podman/bin/podman image list --format '{{ .Digest }}' ghcr.io/almet/dangerzone/dangerzone-testing
Error calling Python override of QThread::run(): Traceback (most recent call last):
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/podman/command/cli_runner.py", line 261, in run_raw
    ret = subprocess.run(
          ^^^^^^^^^^^^^^^
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/capture_output.py", line 89, in patched_subprocess_run
    process = original_subprocess_run(
              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.9/Frameworks/Python.framework/Versions/3.12/lib/python3.12/subprocess.py", line 573, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/opt/podman/bin/podman', 'image', 'list', '--format', '{{ .Digest }}', 'ghcr.io/almet/dangerzone/dangerzone-testing']' returned non-zero exit status 125.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/startup.py", line 232, in run
    if task.should_skip():
       ^^^^^^^^^^^^^^^^^^
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/startup.py", line 166, in should_skip
    return installer.get_installation_strategy() == InstallationStrategy.DO_NOTHING
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/updater/installer.py", line 105, in get_installation_strategy
    podman_images = runtime.list_image_digests()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/container_utils.py", line 244, in list_image_digests
    podman.run(
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/podman/command/command.py", line 80, in run
    return self.runner.run(
           ^^^^^^^^^^^^^^^^
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/podman/command/cli_runner.py", line 217, in run
    return self.run_raw(
           ^^^^^^^^^^^^^
  File "/Users/alexis/dev/fpf/dangerzone/dangerzone/podman/command/cli_runner.py", line 267, in run_raw
    raise errors.CommandError(e) from e
dangerzone.podman.errors.exceptions.CommandError: The Podman process failed with the following error: Command '['/opt/podman/bin/podman', 'image', 'list', '--format', '{{ .Digest }}', 'ghcr.io/almet/dangerzone/dangerzone-testing']' returned non-zero exit status 125.
Stderr: b'Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM\nError: unable to connect to Podman socket: failed to connect: dial tcp 127.0.0.1:64148: connect: connection refused\n'
```

The UI is currently staying at "Checking for updates" indefinitely.

apyrgio commented:
> Maybe we should reset the container_runtime argument to its default value for the 0.10.0 upcoming release?

That's not necessary I believe, because people with this setting on should get the exact same experience as in 0.9.0. If their machine was not started then, Dangerzone would also show an error. So to me, it's just a matter of showing a descriptive error, which is easy to now that we have signals for pretty much everything.

> What's done is not needed anymore. If you have it under your radar, then do as you please ;)

Thanks. I may add a small commit for at least having a proper error screen, and then we'll do the polishing in our UX session.
